### ISSUE TITLE
Calling SDL_Init and SDL_Quit in a thread causes System.AccessViolationException

### ISSUE THREAD

**Issue by ashwinm76** [flibitijibibo/SDL2-CS#87]
If SDL_Init and SDL_Quit are called multiple times in a thread, then they cause a System.AccessViolationException on the SDL_Init call.

To illustrate, I created a simple test console application:

    namespace SDL_test_console
    {
        class Program
        {
            static void Main(string[] args)
            {
                for (int i = 0; i < 10; i++)
                {
                    Console.WriteLine("Count: {0}", i);
                    Thread t = new Thread(new ThreadStart(() =>
                    {
                        SDL.SDL_Init(SDL.SDL_INIT_EVERYTHING);
                        SDL.SDL_Quit();
                    }));
                    t.Start();
                    t.Join();
                }
            }
        }
    }

When I run this. I reliably get the exception on the third call to SDL_Init. 

The same thing occurs in a WPF application: I need to wrap SDL_Init and SDL_Quit in `Dispatcher.Invoke()` to not cause an exception.

Are SDL_Init and SDL_Quit only supposed to be called in the main application thread? I didn't find any mention of this online. Also, can SDL_Init and SDL_Quit be called from different threads?

Thanks.

**Comment by ashwinm76** [comment]
More points to note:
- If instead of SDL_INIT_EVERYTHING I use OR'd flags for all peripherals except SDL_INIT_HAPTIC then the problem disappears.
- After uninstalling a driver for the PS3 move controller on my machine, the problem disappeared, even when using SDL_INIT_EVERYTHING.

So my question now is, could this be a problematic driver issue, or is it not recommended to call SDL_Init and SDL_Quit in a thread other than the main application thread?

**Comment by flibitijibibo** [comment]
Generally SDL_Init and SDL_Quit should only be called once... maybe twice if you call these two in order each time (i.e. the FNA Game calls SDL_Init in the constructor and SDL_Quit in the Dispose routine, and some games that crash will use a smaller Game that prints the Exception trace). Additionally, the joystick subsystem for Win32 in particular needs the [STAThread] attribute since XInput and DirectInput require STA for whatever reason. This is likely why you're seeing crashes, not just because you're constantly (and with _multiple_ threads even!) initing and quitting SDL over and over again, but the subsystems SDL loads also get loaded/unloaded as well (unless you're somehow locking your static init/quits somehow... I really would like to see the actual shipping code for this just out of curiosity).

AccessViolationExceptions with the C# wrappers generally mean something's wrong on your end.

**Comment by ashwinm76** [comment]
Thanks for the reply, that was very helpful! 

My application is a BASIC interpreter with an IDE/debugger that uses SDL for input, video and audio. Because it's an IDE, it needs to start/stop the SDL subsystem every time the user runs/debugs the BASIC program. And since the SDL event loop must not block the IDE's GUI event loop, I put the SDL_Init, SDL_Quit and SDL event polling in a separate thread. So based on your information, I should probably be setting that thread's model to STA manually. Or I should revisit my architecture.

**Comment by ashwinm76** [comment]
Ok, more investigation revealed the following:
- At least for the simple test program I listed above, the thread apartment state is irrelevant; no matter what it's set to, the program crashes.
- Debugging within the SDL2 dll code revealed that the crash occurs when the haptic subsystem tries to free some memory in SDL_Quit. Removing the haptic subsystem from the SDL_Init call makes the crash go away.
- Uninstalling a Windows PS3 Dualshock driver (which seems to install two "phantom" Xinput devices) make the crash go away, even when the haptic subsystem is included in the SDL_Init call.

In conclusion, the problem I'm seeing is probably some weird interaction between SDL's haptic subsystem and the PS3 Dualshock driver. This issue can be closed, since even if this is an SDL problem, it belongs with the SDL2 project and not SDL2-CS. Thanks for your help!
